File: src/main/scala/de/fhopf/akka/sequential/SequentialExecution.scala
package de.fhopf.akka.sequential

import de.fhopf.akka._
import de.fhopf.akka.Execution
import de.fhopf.akka.Executor
import org.apache.lucene.index.IndexWriter

/**
 * @author Florian Hopf, http://www.florian-hopf.de
 * @author Alfredo Serafini, http://www.seralf.it
 */
class SequentialExecution extends Execution {

  override def downloadAndIndex(path: String, writer: IndexWriter) = {

    val pageStore: VisitedPageStore = new VisitedPageStore()
    pageStore.add(path)
    val indexer: Indexer = new IndexerImpl(writer)
    val retriever: PageRetriever = new HtmlParserPageRetriever(path)

    var page = ""
    // TODO: check the correct values here
    while ((page = pageStore.getNext()) != None) {
      val pageContent: PageContent = retriever.fetchPageContent(page)
      pageStore.addAll(pageContent.getLinksToFollow())
      indexer.index(pageContent)
      pageStore.finished(page)
    }

    indexer.commit()
  }

  def main(args: Array[String]) {
    val execution: SequentialExecution = new SequentialExecution()
    val exec: Executor = new Executor(execution)
    exec.execute("http://www.synyx.de/")
  }

}

File: src/main/java/de/fhopf/akka/sequential/SequentialExecution.java
package de.fhopf.akka.sequential;

import de.fhopf.akka.*;
import de.fhopf.akka.Execution;
import de.fhopf.akka.Executor;
import org.apache.lucene.index.IndexWriter;

/**
 * Indexes pages sequentially.
 * @author Florian Hopf, http://www.florian-hopf.de
 */
public class SequentialExecution implements Execution {

    @Override
    public void downloadAndIndex(String path, IndexWriter writer) {
        VisitedPageStore pageStore = new VisitedPageStore();
        pageStore.add(path);
        
        Indexer indexer = new IndexerImpl(writer);
        PageRetriever retriever = new HtmlParserPageRetriever(path);
        
        String page;
        while ((page = pageStore.getNext()) != null) {
            PageContent pageContent = retriever.fetchPageContent(page);
            pageStore.addAll(pageContent.getLinksToFollow());
            indexer.index(pageContent);
            pageStore.finished(page);
        }
        
        indexer.commit();
    }
    
    public static void main(String[] args) {
        SequentialExecution execution = new SequentialExecution();
        Executor exec = new Executor(execution);
        exec.execute("http://www.synyx.de/");
    }
    
}
